Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,516 @@
+
+            case .none:
+                break
+            }
+        }
+
+        private func makeConnection(on eventLoop: EventLoop) -> EventLoopFuture<Connection> {
+            let bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: eventLoop, host: self.key.host, port: self.key.port, configuration: self.configuration) { channel in
+                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration).flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }
+            }
+            let address = HTTPClient.resolveAddress(host: self.key.host, port: self.key.port, proxy: self.configuration.proxy)
+            return bootstrap.connect(host: address.host, port: address.port).map { channel in
+                let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                self.configureCloseCallback(of: connection)
+                connection.isLeased = true
+                return connection
+            }
+        }
+
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    let action = self.lock.withLock { self.state.removeClosedConnection(connection) }
+                    switch action {
+                    case .none:
+                        break
+                    case .removeProvider:
+                        self.parentPool[self.key] = nil
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func syncClose(requiresCleanClose: Bool) throws {
+            let availableConnections = try self.lock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
+                if requiresCleanClose {
+                    guard self.state.leased == 0 else {
+                        throw HTTPClientError.uncleanShutdown
+                    }
+                }
+                return self.state.availableConnections
+            }
+            do {
+                try EventLoopFuture<Void>.andAllComplete(availableConnections.map { $0.channel.close() }, on: self.eventLoop).wait()
+            } catch ChannelError.alreadyClosed {
+                return
+            } catch {
+                throw error
+            }
+            self.lock.withLock { self.state.availableConnections.removeAll() }
+        }
+
+        private struct State {
+            /// The default `EventLoop` to use for this `HTTP1ConnectionProvider`
+            private let defaultEventLoop: EventLoop
```

Comment:
How and when is this different from the parent object's `eventLoop` property?

Reply:
They are the same event loop, maybe I should instead remove the one from the parent and use `state.eventLoop` to avoid property redundancy?